### PR TITLE
AAInstr: implement optimizers

### PR DESCRIPTION
**Pull request by webmiche**
This PR starts implementing the optimizers. In particular, it adds a driver class and the first "optimizer" (the imprecise one).

### CODE REVIEW

File: AAInstrumentation/core.py [hunk 1]
```diff
@@ -117,7 +117,7 @@ def generate_baseline(self):
                 "build",
                 str(self.initial_dir),
                 str(self.benchmark),
-                str(self.instr_path.joinpath("clang")),
+                str(self.instr_path / Path("clang")),
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: Given that `.instr_path` is already a `Path` you can write this instead: `self.instr_path/"clang"`

File: AAInstrumentation/core.py [hunk 2]
```diff
@@ -236,6 +236,23 @@ def run_and_assemble_file(
...
         )
 
+    def run_assemble_and_measure_file(
+        self,
+        file_name: Path,
+        name_prefix: int,
+        index_list: list[int],
+        instrument_recursively=False,
+    ) -> int:
+        """Run a single function, assemble the result, and measure the output size."""
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: > Run a single function

what does this mean?

File: AAInstrumentation/optimization.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    This class implements the imprecise search algorithm. It is used as a
+    baseline for the other optimizers.
+    """
+
+    description = "Imprecise search"
+
+    def optimize(self, file_name: Path, num_candidates: int) -> int:
+        return self.run_assemble_and_measure_file(
+            file_name, 0, list(range(num_candidates))
+        )
```
Comment (thetheodor) [inline AAInstrumentation/optimization.py:73]: Does this really do anything? Other than just checking one query sequence?

**Review by thetheodor — APPROVED** [review]

File: AAInstrumentation/optimization.py (hunk 3 above)
Comment (webmiche) [reply AAInstrumentation/optimization.py:73]: This is an implementation of one of the baselines. If you remember, next to comparing to precise AAInfo, I also compared to fully imprecise info. This instrumentation for the fully imprecise thing is literally just adding all indexes, so I decided to use it as the proof of concept for this first "optimizer".

File: AAInstrumentation/core.py (hunk 2 above)
Comment (webmiche) [reply AAInstrumentation/core.py]: It means I copy pasted this from the old instrumentation. Change to "Run a round of instrumentation, assemble the result, and measure the output size."

**Comment by webmiche** [comment]
@thetheodor let's see if this atually works...
I want to merge this PR first, as the other one is stacked on top of this.

**Comment by webmiche** [comment]
So, I once again request your guidance

File: AAInstrumentation/optimization.py (hunk 3 above)
Comment (thetheodor) [reply AAInstrumentation/optimization.py:73]: The description and class name confused me.

> the imprecise search algorithm

Reading this makes me think that there's some sort of search going on.

**Comment by thetheodor** [comment]
Other than the confusing name of the imprecise, it looks good  to me

File: AAInstrumentation/optimization.py (hunk 3 above)
Comment (webmiche) [reply AAInstrumentation/optimization.py:73]: Good point, changed :+1:

**Pull request merged**